### ISSUE TITLE
Comments on julia-crash-course-STANDARD-02-02-2023

### ISSUE THREAD

**Issue by RechercheStochastique** [JuliaZoid/JuliaCrashCourse-Feedback#2]
Mr. Kilpatrick and Fortman

Thank you for making available the first version of Julia Crash Course STANDARS 02 02 2023. I was looking for such a book about a year ago when I started learning Julia. I am pleased to see this happening now, and wish you much success with this endeavor. 
I have gone through the document and I have comments that I hope you will find helpful. Feel free to ignore any of them, they were made in good faith.
 I have divided my comments into categories and I used “_italics_” in double quotes to indicate sections of the document.
Book structure

1. 	“_Setting up Julia_” should be move into the “_Basics_” section as the first subsection. You need first to discuss the pros and cons of the language.

2.	I would move the “_Why Julia_” section into a foreword section. To me it is not really a “_Basics_” topic. I would make it a lot shorter on some topics and add many more. My point of view is as follow: someone reading the foreword want to know if Julia is the right programming language. Every aspect should be covered shortly. This foreword would fit right after “_How to use this book_“ section. Here’s a list of topics I would like to see in this section. 
a)	I feel that the performance discussion in “_Why Julia_” is too long. We get the point. Dwelling on performance feeling such as startup time or the first plot could be reduced to one or two sentences. I don’t think you should be justifying these. If you want, you could create a final section for future improvements at the end of the book. As the second edition comes, then you simply remove fixed problem and add new area of improvements. The point you want to make here is: if speed is a concern, Julia may be of interest, if not… then Python, R or else is sufficient. 
b)	The REPL, shell mode, Pkg and Help should also be mentioned in a shorter section of “_Why Julia_” (in the foreword). These are good reasons to use Julia but at this step, no need to explain how to start Pkg. You should instead mention the rich universe of packages and explain why this is so great. I am an old C programmer and, unfortunately, this did not exist back then. A clear but short overview of the environment is good at this point to convince someone that Julia is a **modern language and offers a rich working environment**. At this point no need for screenshots.
c)	Pkg is discussed in “_The Package Ecosystem_” in the “_Basics_” section. This is good. The “_Why Julia_” would only mention it and say nice things about the **rich set of packages**. Then, in the “_Basics_” section you should dig into the details.
d)	The help mode is also a good thing to mention in “_Why Julia_”. No need to know that we type ‘?’. We need to understand that Julia provides an **easy way to document everything in the code that will automatically generate help**. 
e)	The Documenter package should also be mentioned there as a plus. Not how-to, but **how easy web pages will be created**. Another reason to like Julia!
f)	“_Intuitive Syntax_”. People coming from other language will be puzzled by the freedom of writing in Julia. I don’t know how to explain this simply. It implies an in-depth discussion on types and multiple dispatch. It should simply mention here why an explanation of the “why” not the “how”. This is in fact a thing I struggled first with. You may leave this as an argument. Suffice to say that I don’t agree and yet, I am a Julia fan.
g)	“_1-based indexing_”. Is that a reason to use Julia? Honestly, I don’t think so. If the section was called “_Julia Peculiarities_”, then a subsection on indexing would be appropriate. Definitively NOT a reason to use Julia. Nobody really cares IMHO.
h)	“_Conditionals and Loops_”. Too early, should be moved to “_Basics_”.
i)	“_The Package Ecosystem_”. Again, this is good but too long for the “_Why Julia_” section. Simply get to the point how this is good. You can mention where to find the package list, and discuss some of them if you which. But not how to used them. Pkg use should be discussed in-depth in the first “Basics” section with REPL, help, etc.
j)	“_Job Opportunities_”. Please remove this.
k)	**Nothing on JIT compilation**? Users will ask: where’s Makefile? I quick discussion on the philosophy behind the creation of Julia would be nice. It could even be right at the beginning of the “_Why Julia_” section.
l)	**LaTeX used in Julia or UTF-8 characters**. As a mathematician I love it. But some may not find it that helpful. However, it makes string manipulation a bit unintuitive at first.
m)	**Testing utilities** are also something I find useful in Julia. Nothing extraordinary but it is integrated into the language. Like the help or the documentation in general. This is a plus.
n)	**Multiple Dispatch!!!!** Newcomers don’t know what it is but have probably heard that this is the paradigm Julia is based on. You need to give it a rapid discussion. Why not OOP? Why is it better? No need to go in details but this is definitively a why?. This brings back to idea of the philosophy behind.
o)	**Jupyter Notebook** is also a why Julia. It should be mentioned as a plus.
3.	“_Julia Community_”. I honestly don’t know where to put this. I fully agree with the content of this section, but where should it go? Is it a topic for a subsection of “_Why Julia_”? Then it should be short. It could refer to more information at the end of the book where this is discussed more in depth with URL to packages, Disclosure, etc. Just an idea.
4)	As mentioned previously, the “_Setting up Julia_” section should be in “_Basics_” section. It could be the first one. Since it involves testing REPL, I suggest that REPL becomes the next section following the setup section. Further, I thing you should explain all the tools at the user’s disposal before digging into the even simple examples: help, pkg, shell. Well, “Hello World!” could be there since it’s the first thing people will be looking for, but not too much syntax yet. Downloading packages, pre-compiling them, etc. should be explained there. This is a bit confusing to newcomers. 
5)	“_Conclusion_” for the “_Setting up Julia_” is redundant. I would remove.
6)	“_Hello World!_” section is way too long for such an easy thing. But this is the good place to talk in-depth about the tools: REPL, Pkg, help, shell. It is currently in section 8 “_The REPL_”. This is too late IMHO. 
7)	To summarize, the “_Basics_” section would be like this:
a)	"_Setting up Julia_", including directory structure and setting up a project. Project.toml and Manifest.toml have to be explained.
b)	"_Tools_": REPL, help (with a quick reference to Documenter to be discussed later), shell, Pkg. An in-depth discussion on pre-compiling, JIT, etc. is needed here. Testing can be deferred to later.
c)	Julia language syntax…. Several sections as you already have planned. You could even separate this from the “Basics” above in a “Julia language” section of its own. Only after item a) and b) above are addressed should we start looking at the syntax.
d)	"_Documenter_"
e)	"_Testing_"
f)	"_Notebook_"
 
Content
1)	“_Variables and Types_” section needs a good discussion on Types. If the goal is to try to convince people that Julia is easy like Python, I think the book will miss its goal. Why Julia if you simply want something easy? The notion of Types in Julia is much more profound than it looks at first and it took me a significant time to understand how rich this is. Something like ::<Complex needs a discussion.
2)	“_Managing Julia Packages in a Notebook_”. Anything related to Jupyter notebook should be a special section. I did a lot of Julia programming without even knowing how to setup a notebook. IMHO it is an advanced topic.
3)	Julia is a complex language, it is not C, and should not be presented as a wannabe Python. Julia is rich and **it will be difficult to keep the book at a reasonable size**. Personally, I would aim for a complete presentation even if the book gets big. A possible approach would be in two steps: 1- get someone able to do something significant with a package without going too much into the details (I wanna plot my dataframe), 2- then continue the book for adults.
4)	The Python package bash in page 77 should be removed. Don’t get my wrong, I dislike Python. But this paragraph is useless to learn Julia and leaves a bad impression. Besides, why all due respect, I don’t think readers care about you experience/mood.
5)	The introductory paragraph of section 9 “_Structs_” is not optimal. Firstly, structs in Julia are not comparable to C structures, they are far better. Also, instead of comparing structs with objects, the point should be made about the fact that they have constructors and integrate into multiple dispatch. This is, in fact, superior to objects. There have been several discussions in Disclosure about Types that have opened my eyes on the power of Julia. I honestly don’t know how to present this simply, but the examples provided in these discussions were very helpful.
Presentation
Some of the screenshots are not very inviting. But I presume this will be redone as one of the last steps before publishing.
Pictures are ugly (IMHO). Do you really need to put them? The book will be thick enough.

As a final note, I would encourage you to be more proactive with your supporters like me. The fact that you were late to deliver the first draft is understandable. But as a project manager I can tell you from experience that it is better to manage expectations rapidly, and announce delays when you know they are bound to happen. At one point I was thinking that I was the victim of a scam.

I hope you found my comments helpful. And I look forward to see the next version of the document.

Denys Gagnon
Montréal
Feb 9 2023
